70_FLIPFluids/app_resources/compute/pressureSolver.comp.hlsl

```diff
@@ -0,0 +1,120 @@
+#include "../common.hlsl"
+#include "../gridUtils.hlsl"
+#include "../cellUtils.hlsl"
+#include "../descriptor_bindings.hlsl"
+
+struct SPressureSolverParams
+{
+    float4 coeff1;
+    float4 coeff2;
+};
+
+[[vk::binding(b_psGridData, s_ps)]]
+cbuffer GridData
+{
+    SGridData gridData;
+};
+
+[[vk::binding(b_psParams, s_ps)]]
+cbuffer PressureSolverParams
+{
+    SPressureSolverParams params;
+};
+
+[[vk::binding(b_psCMBuffer, s_ps)]] RWStructuredBuffer<uint> cellMaterialBuffer;
+[[vk::binding(b_psVelBuffer, s_ps)]] RWTexture3D<float4> velocityFieldBuffer;
+[[vk::binding(b_psDivBuffer, s_ps)]] RWStructuredBuffer<float> divergenceBuffer;
+[[vk::binding(b_psPresInBuffer, s_ps)]] RWStructuredBuffer<float> pressureInBuffer;
+[[vk::binding(b_psPresOutBuffer, s_ps)]] RWStructuredBuffer<float> pressureOutBuffer;
+
+[numthreads(WorkgroupSize, 1, 1)]
+void calculateNegativeDivergence(uint32_t3 ID : SV_DispatchThreadID)
+{
+    uint cid = ID.x;
+    int3 cellIdx = flatIdxToCellIdx(cid, gridData.gridSize);
+
+    float3 param = (float3)gridData.gridInvCellSize;
+    float3 velocity = velocityFieldBuffer[cellIdx].xyz;
+
+    float divergence = 0;
+    if (isFluidCell(getCellMaterial(cellMaterialBuffer[cid])))
+    {
+        int3 cell_xn = cellIdx + int3(1, 0, 0);
+        uint cid_xn = cellIdxToFlatIdx(cell_xn, gridData.gridSize);
+        divergence += param.x * ((cell_xn.x < gridData.gridSize.x ? velocityFieldBuffer[cell_xn].x : 0.0f) - velocity.x);
+
+        int3 cell_yn = cellIdx + int3(0, 1, 0);
+        uint cid_yn = cellIdxToFlatIdx(cell_yn, gridData.gridSize);
+        divergence += param.y * ((cell_yn.y < gridData.gridSize.y ? velocityFieldBuffer[cell_yn].y : 0.0f) - velocity.y);
+
+        int3 cell_zn = cellIdx + int3(0, 0, 1);
+        uint cid_zn = cellIdxToFlatIdx(cell_zn, gridData.gridSize);
+        divergence += param.z * ((cell_zn.z < gridData.gridSize.z ? velocityFieldBuffer[cell_zn].z : 0.0f) - velocity.z);
+    }
+
+    divergenceBuffer[cid] = divergence;
+}
```

Comment:
this can be kernel fused with the updateVelocity and diffusion pipeline if you preaload a 1 cell border around 8x8x8 into shared memory and do some redundant work